### PR TITLE
Implement log partitioning

### PR DESCRIPTION
**Pull request by lgolding**
This new partitioning API is intended for use in filing work items.

Since I have logic to avoid excessive cloning (you can see this by looking at where `deepClone` is used), we might not need the `deepClone` flag at all: it might be ok to deep clone always.

Note:
- Testing this feature required an enhancement to `FileDiffingUnitTests` so a single test can produce multiple output files. (A single test can now also consume multiple input files.)

### CODE REVIEW

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+        {
+            SarifLog partitionLog;
+            if (deepClone)
+            {
+                // Save time and space by not cloning the runs unless and until necessary. We will
+                // only need to clone the runs that have results in this partition.
+                IList<Run> originalRuns = originalLog.Runs;
+                originalLog.Runs = null;
+
+                partitionLog = originalLog.DeepClone();
```
Comment (ScottLouvau) [inline src/Sarif/Visitors/PartitioningVisitor.cs]: Is this actually different than new SarifLog(originalLog)? Are there any properties outside the Runs which need DeepCloning?

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 2]
```diff
@@ -0,0 +1,454 @@
...
+                    // certain run-level collections such as Run.Artifacts, have been replaced.
+                    Run originalRun = originalLog.Runs[iOriginalRun];
+                    Run partitionRun;
+                    if (deepClone)
+                    {
+                        // Save time and space by only cloning the necessary results and associated
+                        // collection elements. We already cloned the relevant results in VisitResult.
+                        IList <Result> originalResults = originalRun.Results;
+                        IList<Artifact> originalArtifacts = originalRun.Artifacts;
+                        originalRun.Results = null;
```
Comment (ScottLouvau) [inline src/Sarif/Visitors/PartitioningVisitor.cs]: I think a function wrapping the "DeepCloneExceptXAndY" functionality would help. It took me a moment to figure out that's what you're doing (storing, clearing, cloning, and then resetting the references you don't want cloned)

File: src/Test.FunctionalTests.Sarif/Partitioning/PartitionTests.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        {
+            Partition_ByRuleId_ProducesOneLogFilePerRule(deepClone: true);
+        }
+
+        [Fact]
+        public void Partition_ByRuleId_WithShallowCopy_ProducesOneLogFilePerRule()
+        {
+            Partition_ByRuleId_ProducesOneLogFilePerRule(deepClone: false);
+        }
+
```
Comment (ScottLouvau) [inline src/Test.FunctionalTests.Sarif/Partitioning/PartitionTests.cs]: If you have a sample with more than one run, you should check that Runs are excluded altogether if none of the Run Results is in a given partition. (If not already covered by the above)

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+        /// result belongs to, or null if the result should be discarded (not placed in any
+        /// partition).
+        /// </summary>
+        /// <param name="result">
+        /// The result to be assigned to a partition.
+        /// </param>
+        /// <typeparam name="T">
+        /// The type of the object returned from the partition function. It must be a reference
+        /// type so that null is a valid value. It must override bool Equals(T other) so that
+        /// two Ts can compare equal even if they are not reference equal.
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Interesting point. According to [StackOverflow](https://stackoverflow.com/questions/15889987/when-to-specify-constraint-t-iequatablet-even-though-it-is-not-strictly-re), it doesn't actually add anything because all types are convertible to `System.Object`, which implements `IEquatable<T>`.

**BUT** The poster goes on to argue that it makes for good documentation, telling the user that the objects are going to be compared, so perhaps he should give some thought to how the comparison would work.

On the other other hand, my comment already does that.

TL;DR: I'll add it.

---
In reply to: [357427829](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357427829) [](ancestors = 357427829)

File: src/Sarif/Visitors/PartitioningVisitor.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Yes, `sarifLog.inlineExternalProperties`.

---
In reply to: [357431410](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357431410) [](ancestors = 357431410)

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+
+        public override Result VisitResult(Result node)
+        {
+            inResult = true;
+            currentPartitionValue = partitionFunction(node);
+            if (currentPartitionValue != null)
+            {
+                if (!partitionValues.Contains(currentPartitionValue))
+                {
+                    partitionValues.Add(currentPartitionValue);
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Yes, thanks. I did that elsewhere, missed it here.

---
In reply to: [357429173](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357429173) [](ancestors = 357429173)

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+        /// <summary>
+        /// Returns a mapping from partition value to the partition log for that value.
+        /// </summary>
+        /// <returns></returns>
+        public Dictionary<T, SarifLog> GetPartitionLogs()
+        {
+            if (originalLog == null)
+            {
+                throw new InvalidOperationException(
+                    "You must call " + nameof(VisitSarifLog) +
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Done.

---
In reply to: [357428813](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357428813) [](ancestors = 357428813)

File: src/Sarif/Visitors/PartitioningVisitor.cs (hunk 2 above)
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: It's a nice idea but the function name (which would now be `DeepCloneExceptResultsAndArtifacts`) will become unwieldy once it's `DeepCloneExceptResultArtifactsGraphsLogicalLocationsThreadFlowLocationsWebRequestsAndWebResponses` :-)

---
In reply to: [357431653](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357431653) [](ancestors = 357431653)

File: src/Test.FunctionalTests.Sarif/Partitioning/PartitionTests.cs (hunk 3 above)
Comment (lgolding) [reply src/Test.FunctionalTests.Sarif/Partitioning/PartitionTests.cs]: The test cases do already cover that. The input file Partition.sarif has two runs. Note that the output file TST0002.sarif contains only one run (corresponding to the second of the two runs in the input file), because only that run has any results with `ruleId` `TST0002`.

---
In reply to: [357432826](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357432826) [](ancestors = 357432826)

File: src/Sarif/Visitors/PartitioningVisitor.cs [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        // True if we are currently processing a result, otherwise false.
+        private bool inResult;
+
+        // The set of results in each partition in each original run. That is,
+        //
+        //     partitionedResultDictionaries[runIndex][partitionValue]
+        //
+        // contains the list of results in the run at runIndex for whose partition value is
+        // partitionValue.
+        private List<Dictionary<T, List<Result>>> partitionResultDictionaries;
```
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: It's a nice idea, it wouldn't be that hard, and I have a good set of test to backstop me. I'll do it.

---
In reply to: [357428561](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357428561) [](ancestors = 357428561)

File: src/Sarif/Visitors/PartitioningVisitor.cs (hunk 2 above)
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Oh, I suppose `DeepCloneExceptRunLevelCollections` would be ok. Let me think about it.

---
In reply to: [357441324](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357441324) [](ancestors = 357441324,357431653)

Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: No, the comment already says we're "cloning only the necessary results and associated collection elements." I'm going to leave this.

---
In reply to: [357442600](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357442600) [](ancestors = 357442600,357441324,357431653)

File: src/Sarif/Visitors/PartitioningVisitor.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: Done. Really quite a small change, and a nice one.

---
In reply to: [357442354](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357442354) [](ancestors = 357442354,357428561)

File: src/Sarif/Visitors/PartitioningVisitor.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif/Visitors/PartitioningVisitor.cs]: (Even if there weren't any such properties, it's good future-proofing against the day when there might be some.)

---
In reply to: [357440439](https://github.com/microsoft/sarif-sdk/pull/1719#discussion_r357440439) [](ancestors = 357440439,357431410)
